src/include/sof/audio/crossover/crossover.h

```diff
@@ -142,14 +143,101 @@ static inline crossover_split crossover_find_split_func(int32_t num_sinks)
 	return crossover_split_fnmap[num_sinks - CROSSOVER_2WAY_NUM_SINKS];
 }
 
+static int32_t iir_df2t_2stage_biquads(struct iir_state_df2t *iir, int32_t x)
+{
+	int32_t in;
+	int32_t tmp;
+	int32_t tmp_out;
+	int64_t acc;
+	int32_t out = 0;
+	int d = 0; /* Index to delays */
+	int c = 0; /* Index to coefficient a2 */
+
+	/* Bypass is set with number of biquads set to zero. */
+	if (!iir->biquads)
+		return x;
+
+	/* Coefficients order in coef[] is {a2, a1, b2, b1, b0, shift, gain} */
+
+	/* First-stage biquad filter */
+
+	/* Compute output: Delay is Q4.60
+	 * Coef(Q2.30) x Input(Q1.31) = Q3.61 shift to Q4.60 for accumulating
+	 * Then saturate to Q4.28
+	 */
+	in = x;
+	acc = iir->delay[d];
+	acc += Q_MULTSR_32X32((int64_t)iir->coef[c + 4], in, 30, 31, 60); /* b0 */
+	tmp = sat_int32((int32_t)Q_SHIFT_RND(acc, 60, 28));
```

Comment:
I think here the (int32_t) cast for Q_SHIFT_RND() is not correct, you should leave it 64 bits.

I do agree this saturation operation is needed that I've missed to add to generic-C version but the fact that this overflow triggered for you means that your IIR scaling is incorrect. The saturation of recursive filter state variables can  trigger instability (oscillation). I will do similar change to IIR but I really recommend to change your biquads scaling. 

I'll do also the Q-format change of tmp to make it robust but I assume IIR THD+N performance takes a bit of hit from it, especially with proper biquads scaling applied.

Reply:
This function is no more applicable.